Lib/ntpath.py

```diff
@@ -167,40 +171,31 @@ def splitdrive(p):
 
     """
     p = os.fspath(p)
-    if len(p) >= 2:
-        if isinstance(p, bytes):
-            sep = b'\\'
-            altsep = b'/'
-            colon = b':'
-            unc_prefix = b'\\\\?\\UNC'
-        else:
-            sep = '\\'
-            altsep = '/'
-            colon = ':'
-            unc_prefix = '\\\\?\\UNC'
-        normp = p.replace(altsep, sep)
-        if (normp[0:2] == sep*2) and (normp[2:3] != sep):
-            # is a UNC path:
-            # vvvvvvvvvvvvvvvvvvvv drive letter or UNC path
-            # \\machine\mountpoint\directory\etc\...
-            #           directory ^^^^^^^^^^^^^^^
-            if normp[:8].upper().rstrip(sep) == unc_prefix:
-                start = 8
-            else:
-                start = 2
-            index = normp.find(sep, start)
-            if index == -1:
-                return p[:0], p
-            index2 = normp.find(sep, index + 1)
-            # a UNC path can't have two slashes in a row
-            # (after the initial two)
-            if index2 == index + 1:
-                return p[:0], p
-            if index2 == -1:
-                index2 = len(p)
-            return p[:index2], p[index2:]
-        if normp[1:2] == colon:
-            return p[:2], p[2:]
+    if isinstance(p, bytes):
+        sep = b'\\'
+        altsep = b'/'
+        colon = b':'
+        unc_prefix = b'\\\\?\\UNC\\'
+    else:
+        sep = '\\'
+        altsep = '/'
+        colon = ':'
+        unc_prefix = '\\\\?\\UNC\\'
+    normp = p.replace(altsep, sep)
+    if normp[1:2] == colon and normp[:1] != sep:
+        # Drive-letter logical drives, e.g. X:
+        return p[:2], p[2:]
+    if normp[:2] == sep * 2:
+        # UNC drives, e.g. \\server\share or \\?\UNC\server\share
+        # Device drives, e.g. \\.\device or \\?\device
+        start = 8 if normp[:8].upper() == unc_prefix else 2
+        index = normp.find(sep, start)
+        if index == -1:
+            return p, p[:0]
+        index = normp.find(sep, index + 1)
+        if index == -1:
+            return p, p[:0]
+        return p[:index], p[index:]
```

Comment:
If a share component is empty, the API considers the root to be the slash after the server component. For example:

```pycon
>>> nt._path_splitroot('//x//y')
('//x/', '/y')
>>> nt._getfullpathname('//x//y')
'\\\\x\\y'
```

Invalid "//x//y" thus gets normalized as the valid share path "\\\\x\\y". Also, in such cases components after the server name can be consumed by ".." components. For example:

```pycon
>>> nt._getfullpathname('//x//y/..')
'\\\\x\\'
```

Another quirk is that the "." and ".." component names (i.e. `curdir` and `pardir`) aren't supported as share names in `GetFullPathNameW()`. For example: 

>>> nt._getfullpathname('//x/./..')
'\\\\x\\'
>>> nt._getfullpathname('//x/../..')
'\\\\x\\'

I suggest that `splitdrive()` should conform with the API in such cases.

```suggestion
    if isinstance(p, bytes):
        sep = b'\\'
        altsep = b'/'
        curdir = b'.'
        pardir = b'..'
        colon = b':'
        unc_prefix = b'\\\\'
        unc_device_prefix = b'\\\\?\\UNC\\'
    else:
        sep = '\\'
        altsep = '/'
        curdir = '.'
        pardir = '..'
        colon = ':'
        unc_prefix = '\\\\'
        unc_device_prefix = '\\\\?\\UNC\\'
    normp = p.replace(altsep, sep)
    if normp[1:2] == colon and normp[:1] != sep:
        # Drive-letter drives, e.g. X:
        return p[:2], p[2:]
    if normp[:2] == unc_prefix:
        # UNC drives, e.g. \\server\share or \\?\UNC\server\share
        # Device drives, e.g. \\.\device or \\?\device
        start = 8 if normp[:8].upper() == unc_device_prefix else 2
        index1 = normp.find(sep, start)
        if index1 == -1:
            return p, p[:0]
        index2 = normp.find(sep, index1 + 1)
        if index2 == -1:
            index2 = len(p)
        if p[index1 + 1:index2] in (p[:0], curdir, pardir):
            return p[:index1], p[index1:]
        return p[:index2], p[index2:]
```